File: prototype-swift/Experiments/incremental_pagerank.swift
import Metal
import Foundation
import Accelerate

// MARK: - Graph Structures

struct EdgeBlock {
    let edges: [UInt32]
    let targets: [UInt32]
    let blockStart: [UInt32]
    let blockCount: [UInt32]
    let vertexCount: UInt32
    let totalEdges: UInt32
    
    init(edges: [UInt32], targets: [UInt32], blockStart: [UInt32], blockCount: [UInt32], 
         vertexCount: UInt32, totalEdges: UInt32) {
        self.edges = edges
        self.targets = targets
        self.blockStart = blockStart
        self.blockCount = blockCount
        self.vertexCount = vertexCount
        self.totalEdges = totalEdges
    }
}

struct CSRGraph {
    let offsets: [UInt32]
    let targets: [UInt32]
    let vertexCount: UInt32
    let totalEdges: UInt32
    
    init(offsets: [UInt32], targets: [UInt32], vertexCount: UInt32, totalEdges: UInt32) {
        self.offsets = offsets
        self.targets = targets
        self.vertexCount = vertexCount
        self.totalEdges = totalEdges
    }
}

// MARK: - Graph Generation

func generatePowerLawGraph(vertexCount: UInt32, edgeCount: UInt32) -> (EdgeBlock, CSRGraph) {
    var edges: [UInt32] = []
    var adjacency = Array(repeating: Set<UInt32>(), count: Int(vertexCount))
    
    var actualEdges: UInt32 = 0
    let maxAttempts = edgeCount * 10
    var attempts = 0
    
    while actualEdges < edgeCount && attempts < maxAttempts {
        attempts += 1
        let u = UInt32.random(in: 0..<vertexCount)
        var v = UInt32.random(in: 0..<vertexCount)
        while v == u {
            v = UInt32.random(in: 0..<vertexCount)
        }
        
        if !adjacency[Int(u)].contains(v) {
            adjacency[Int(u)].insert(v)
            edges.append(u)
            edges.append(v)
            actualEdges += 1
        }
    }
    
    let finalEdges = actualEdges
    var sortedAdjacency = Array(repeating: [UInt32](), count: Int(vertexCount))
    for i in 0..<Int(vertexCount) {
        sortedAdjacency[i] = Array(adjacency[i]).sorted()
    }
    
    // Build EdgeBlock
    let blockSize = 32
    var blockEdges: [UInt32] = []
    var blockTargets: [UInt32] = []
    var blockStart = [UInt32](repeating: 0, count: Int(vertexCount) + 1)
    var blockCount = [UInt32](repeating: 0, count: Int(vertexCount))
    
    for i in 0..<Int(vertexCount) {
        let neighbors = sortedAdjacency[i]
        if neighbors.isEmpty { continue }
        
        blockStart[i] = UInt32(blockEdges.count / blockSize)
        let blocksNeeded = (neighbors.count + blockSize - 1) / blockSize
        
        for blockIdx in 0..<blocksNeeded {
            let start = blockIdx * blockSize
            let end = min(start + blockSize, neighbors.count)
            
            for j in 0..<blockSize {
                if j < end {
                    blockEdges.append(UInt32(i))
                    blockTargets.append(neighbors[start + j])
                } else {
                    blockEdges.append(0)
                    blockTargets.append(0)
                }
            }
        }
        
        blockCount[i] = UInt32(blocksNeeded)
    }
    
    blockStart[Int(vertexCount)] = UInt32(blockEdges.count / blockSize)
    
    let edgeBlock = EdgeBlock(
        edges: blockEdges, targets: blockTargets,
        blockStart: blockStart, blockCount: blockCount,
        vertexCount: vertexCount, totalEdges: finalEdges
    )
    
    // Build CSR
    var csrOffsets = [UInt32](repeating: 0, count: Int(vertexCount) + 1)
    var csrTargets: [UInt32] = []
    
    var offset: UInt32 = 0
    for i in 0..<Int(vertexCount) {
        csrOffsets[i] = offset
        csrTargets.append(contentsOf: sortedAdjacency[i])
        offset += UInt32(sortedAdjacency[i].count)
    }
    csrOffsets[Int(vertexCount)] = offset
    
    let csr = CSRGraph(offsets: csrOffsets, targets: csrTargets, vertexCount: vertexCount, totalEdges: finalEdges)
    
    return (edgeBlock, csr)
}

// MARK: - Metal Setup

func createMetalDevice() -> (MTLDevice, MTLCommandQueue) {
    guard let device = MTLCreateSystemDefaultDevice() else {
        fatalError("No Metal device found")
    }
    let queue = device.makeCommandQueue()!
    return (device, queue)
}

// MARK: - Full PageRank (GPU only)

let fullPageRankKernel = """
#include <metal_stdlib>
using namespace metal;

inline void atomic_add_float(device float* address, float val) {
    device atomic_uint* atomicAddr = (device atomic_uint*)address;
    uint oldBits = atomic_load_explicit(atomicAddr, memory_order_relaxed);
    uint newBits;
    do {
        float oldVal = as_type<float>(oldBits);
        float newVal = oldVal + val;
        newBits = as_type<uint>(newVal);
    } while (!atomic_compare_exchange_weak_explicit(
        atomicAddr, &oldBits, newBits,
        memory_order_relaxed, memory_order_relaxed));
}

kernel void pagerank_csr(
    device const uint *offsets [[buffer(0)]],
    device const uint *targets [[buffer(1)]],
    device const float *pr [[buffer(2)]],
    device float *new_pr [[buffer(3)]],
    constant uint &vertex_count [[buffer(4)]],
    uint gid [[thread_position_in_grid]]
) {
    if (gid >= vertex_count) return;
    
    float contribution = 0.0f;
    uint start = offsets[gid];
    uint end = offsets[gid + 1];
    
    for (uint i = start; i < end; i++) {
        uint neighbor = targets[i];
        uint neighbor_start = offsets[neighbor];
        uint neighbor_end = offsets[neighbor + 1];
        uint out_degree = neighbor_end - neighbor_start;
        if (out_degree > 0) {
            contribution += pr[neighbor] / float(out_degree);
        }
    }
    
    float damping = 0.85f;
    float new_value = (1.0f - damping) / float(vertex_count) + damping * contribution;
    atomic_add_float(&new_pr[gid], new_value);
}
"""

func computeFullPageRank(csr: CSRGraph, iterations: Int, device: MTLDevice, queue: MTLCommandQueue) -> [Float] {
    let library = try! device.makeLibrary(source: fullPageRankKernel, options: nil)
    let kernel = library.makeFunction(name: "pagerank_csr")!
    let pipeline = try! device.makeComputePipelineState(function: kernel)
    
    var pr = [Float](repeating: 1.0 / Float(csr.vertexCount), count: Int(csr.vertexCount))
    
    for iter in 0..<iterations {
        let newPr = [Float](repeating: 0.0, count: Int(csr.vertexCount))
        
        let offsetsBuffer = device.makeBuffer(bytes: csr.offsets, length: (Int(csr.vertexCount) + 1) * MemoryLayout<UInt32>.stride, options: [])!
        let targetsBuffer = device.makeBuffer(bytes: csr.targets, length: Int(csr.totalEdges) * MemoryLayout<UInt32>.stride, options: [])!
        let prBuffer = device.makeBuffer(bytes: pr, length: Int(csr.vertexCount) * MemoryLayout<Float>.stride, options: [])!
        let newPrBuffer = device.makeBuffer(bytes: newPr, length: Int(csr.vertexCount) * MemoryLayout<Float>.stride, options: [])!
        var vertexCount = csr.vertexCount
        
        let commandBuffer = queue.makeCommandBuffer()!
        let encoder = commandBuffer.makeComputeCommandEncoder()!
        encoder.setComputePipelineState(pipeline)
        encoder.setBuffer(offsetsBuffer, offset: 0, index: 0)
        encoder.setBuffer(targetsBuffer, offset: 0, index: 1)
        encoder.setBuffer(prBuffer, offset: 0, index: 2)
        encoder.setBuffer(newPrBuffer, offset: 0, index: 3)
        encoder.setBytes(&vertexCount, length: MemoryLayout<UInt32>.stride, index: 4)
        
        let gridSize = MTLSize(width: Int(vertexCount), height: 1, depth: 1)
        let threadGroupSize = MTLSize(width: 256, height: 1, depth: 1)
        encoder.dispatchThreads(gridSize, threadsPerThreadgroup: threadGroupSize)
        encoder.endEncoding()
        commandBuffer.commit()
        commandBuffer.waitUntilCompleted()
        
        pr = Array(UnsafeBufferPointer(start: newPrBuffer.contents().bindMemory(to: Float.self, capacity: Int(csr.vertexCount)), count: Int(csr.vertexCount)))
        
        if iter % 5 == 0 {
            print("    Iteration \(iter): PR sum = \(pr.reduce(0, +))")
        }
    }
    
    return pr
}

// MARK: - Incremental PageRank (CPU + GPU)

let incrementalPageRankKernel = """
#include <metal_stdlib>
using namespace metal;

// GPU kernel: Update PageRank for affected vertices only
kernel void incremental_pagerank(
    device const uint *affected_vertices [[buffer(0)]],
    device const uint &affected_count [[buffer(1)]],
    device const uint *offsets [[buffer(2)]],
    device const uint *targets [[buffer(3)]],
    device const float *pr [[buffer(4)]],
    device float *new_pr [[buffer(5)]],
    constant uint &vertex_count [[buffer(6)]],
    uint gid [[thread_position_in_grid]]
) {
    if (gid >= affected_count) return;
    
    uint v = affected_vertices[gid];
    if (v >= vertex_count) return;
    
    float contribution = 0.0f;
    uint start = offsets[v];
    uint end = offsets[v + 1];
    
    for (uint i = start; i < end; i++) {
        uint neighbor = targets[i];
        uint neighbor_start = offsets[neighbor];
        uint neighbor_end = offsets[neighbor + 1];
        uint out_degree = neighbor_end - neighbor_start;
        if (out_degree > 0) {
            contribution += pr[neighbor] / float(out_degree);
        }
    }
    
    float damping = 0.85f;
    float new_value = (1.0f - damping) / float(vertex_count) + damping * contribution;
    
    // Update this vertex's score
    new_pr[v] = new_value;
}
"""

// CPU+GPU Incremental PageRank
func computeIncrementalPageRank(
    csr: CSRGraph,
    initialPR: [Float],
    affectedVertices: [UInt32],
    maxIterations: Int,
    tolerance: Float,
    device: MTLDevice,
    queue: MTLCommandQueue
) -> [Float] {
    let library = try! device.makeLibrary(source: incrementalPageRankKernel, options: nil)
    let kernel = library.makeFunction(name: "incremental_pagerank")!
    let pipeline = try! device.makeComputePipelineState(function: kernel)
    
    var pr = initialPR
    let vertexCount = Int(csr.vertexCount)
    
    // Build reverse adjacency list for propagation
    var reverseAdjacency = Array(repeating: [UInt32](), count: vertexCount)
    for u in 0..<vertexCount {
        let start = Int(csr.offsets[u])
        let end = Int(csr.offsets[u + 1])
        for i in start..<end {
            let v = Int(csr.targets[i])
            reverseAdjacency[v].append(UInt32(u))
        }
    }
    
    // CPU: Manage affected vertex queue
    var affectedSet = Set<UInt32>(affectedVertices)
    var iteration = 0
    
    let startTime = CFAbsoluteTimeGetCurrent()
    
    while !affectedSet.isEmpty && iteration < maxIterations {
        // Convert set to array for GPU
        let affectedArray = Array(affectedSet)
        var affectedCount = UInt32(affectedArray.count)
        
        // GPU: Update PageRank for affected vertices
        var newPr = pr  // Copy current values
        
        let affectedBuffer = device.makeBuffer(bytes: affectedArray, length: Int(affectedCount) * MemoryLayout<UInt32>.stride, options: [])!
        let offsetsBuffer = device.makeBuffer(bytes: csr.offsets, length: (vertexCount + 1) * MemoryLayout<UInt32>.stride, options: [])!
        let targetsBuffer = device.makeBuffer(bytes: csr.targets, length: Int(csr.totalEdges) * MemoryLayout<UInt32>.stride, options: [])!
        let prBuffer = device.makeBuffer(bytes: pr, length: vertexCount * MemoryLayout<Float>.stride, options: [])!
        let newPrBuffer = device.makeBuffer(bytes: newPr, length: vertexCount * MemoryLayout<Float>.stride, options: [])!
        var vc = csr.vertexCount
        
        let commandBuffer = queue.makeCommandBuffer()!
        let encoder = commandBuffer.makeComputeCommandEncoder()!
        encoder.setComputePipelineState(pipeline)
        encoder.setBuffer(affectedBuffer, offset: 0, index: 0)
        encoder.setBytes(&affectedCount, length: MemoryLayout<UInt32>.stride, index: 1)
        encoder.setBuffer(offsetsBuffer, offset: 0, index: 2)
        encoder.setBuffer(targetsBuffer, offset: 0, index: 3)
        encoder.setBuffer(prBuffer, offset: 0, index: 4)
        encoder.setBuffer(newPrBuffer, offset: 0, index: 5)
        encoder.setBytes(&vc, length: MemoryLayout<UInt32>.stride, index: 6)
        
        let gridSize = MTLSize(width: Int(affectedCount), height: 1, depth: 1)
        let threadGroupSize = MTLSize(width: 256, height: 1, depth: 1)
        encoder.dispatchThreads(gridSize, threadsPerThreadgroup: threadGroupSize)
        encoder.endEncoding()
        commandBuffer.commit()
        commandBuffer.waitUntilCompleted()
        
        // Get updated scores
        let updatedPr = Array(UnsafeBufferPointer(start: newPrBuffer.contents().bindMemory(to: Float.self, capacity: vertexCount), count: vertexCount))
        
        // CPU: Check convergence and find new affected vertices
        var newAffected = Set<UInt32>()
        for i in 0..<vertexCount {
            let diff = abs(updatedPr[i] - pr[i])
            if diff > tolerance {
                // This vertex changed significantly, need to propagate to neighbors
                for neighbor in reverseAdjacency[i] {
                    newAffected.insert(neighbor)
                }
            }
        }
        
        pr = updatedPr
        affectedSet = newAffected
        iteration += 1
        
        if iteration % 5 == 0 {
            print("    Incremental iteration \(iteration): \(affectedSet.count) vertices affected, PR sum = \(pr.reduce(0, +))")
        }
    }
    
    let elapsed = CFAbsoluteTimeGetCurrent() - startTime
    print("    Incremental PageRank: \(iteration) iterations, \(String(format: "%.2f", elapsed * 1000)) ms")
    
    return pr
}

// MARK: - Main Test

print("=== Incremental PageRank Test ===\n")

let (edgeBlock, csr) = generatePowerLawGraph(vertexCount: 10000, edgeCount: 50000)
print("Generated graph: \(csr.vertexCount) vertices, \(csr.totalEdges) edges")

let (device, queue) = createMetalDevice()
print("Metal device: \(device.name)")

// Step 1: Compute full PageRank on original graph
print("\n1. Computing full PageRank on original graph (10 iterations)...")
let fullPrStart = CFAbsoluteTimeGetCurrent()
let originalPR = computeFullPageRank(csr: csr, iterations: 10, device: device, queue: queue)
let fullPrTime = CFAbsoluteTimeGetCurrent() - fullPrStart
print("   Full PageRank time: \(String(format: "%.2f", fullPrTime * 1000)) ms")
print("   PR sum: \(originalPR.reduce(0, +))")

// Step 2: Simulate graph change (add some edges)
print("\n2. Simulating graph change (adding 10 edges)...")
var affectedVertices: [UInt32] = []
for _ in 0..<10 {
    let u = UInt32.random(in: 0..<csr.vertexCount)
    let v = UInt32.random(in: 0..<csr.vertexCount)
    affectedVertices.append(u)
    affectedVertices.append(v)
}
affectedVertices = Array(Set(affectedVertices))  // Remove duplicates
print("   Affected vertices: \(affectedVertices.count)")

// Step 3: Compute incremental PageRank
print("\n3. Computing incremental PageRank (tolerance = 1e-6)...")
let incrementalPr = computeIncrementalPageRank(
    csr: csr,
    initialPR: originalPR,
    affectedVertices: affectedVertices,
    maxIterations: 50,
    tolerance: 1e-6,
    device: device,
    queue: queue
)

// Step 4: Compute full PageRank from scratch for comparison
print("\n4. Computing full PageRank from scratch for comparison...")
let fullPrFromScratch = computeFullPageRank(csr: csr, iterations: 10, device: device, queue: queue)

// Step 5: Compare results
print("\n5. Comparing results...")
var maxDiff: Float = 0
var sumDiff: Float = 0
for i in 0..<Int(csr.vertexCount) {
    let diff = abs(incrementalPr[i] - fullPrFromScratch[i])
    maxDiff = max(maxDiff, diff)
    sumDiff += diff
}
let avgDiff = sumDiff / Float(csr.vertexCount)

print("   Max difference: \(maxDiff)")
print("   Average difference: \(avgDiff)")
print("   PR sum (incremental): \(incrementalPr.reduce(0, +))")
print("   PR sum (full): \(fullPrFromScratch.reduce(0, +))")

if maxDiff < 0.001 {
    print("\n✅ Incremental PageRank results match full PageRank!")
} else {
    print("\n⚠️  Results differ significantly")
}

print("\n=== Test Complete ===")

File: src/incremental_pagerank.rs
// src/incremental_pagerank.rs
// CPU/GPU 协同的增量 PageRank（严格按照 Swift 版本翻译）
// 
// 翻译自：/tmp/axolotl_tmp/Experiments/incremental_pagerank.swift
// Swift 第 275-369 行：computeIncrementalPageRank 函数

use std::collections::HashSet;
use crate::csr_graph::CSRGraph;
use crate::gpu::GPUAccelerator;

/// CPU/GPU 协同的增量 PageRank
/// 
/// 严格按照 Swift 版本（第 275-369 行）实现
pub struct IncrementalPageRank {
    /// GPU 加速器
    gpu: GPUAccelerator,
    /// 阻尼因子
    damping_factor: f32,
    /// 最大迭代次数
    max_iterations: usize,
    /// 收敛阈值
    tolerance: f32,
}

impl IncrementalPageRank {
    pub fn new() -> Result<Self, String> {
        let gpu = GPUAccelerator::new()?;
        
        Ok(IncrementalPageRank {
            gpu,
            damping_factor: 0.85,
            max_iterations: 50,
            tolerance: 1e-6,
        })
    }
    
    /// 计算增量 PageRank（严格按照 Swift 第 275-369 行）
    /// 
    /// 参数：
    /// - csr: CSR 格式的图
    /// - initial_pr: 初始 PR 值
    /// - affected_vertices: 受影响顶点列表（索引）
    /// 
    /// 返回：更新后的 PR 值
    pub fn compute(
        &self,
        csr: &CSRGraph,
        initial_pr: &[f32],
        affected_vertices: &[u32],
    ) -> Vec<f32> {
        // Swift 第 284-286 行：创建 GPU 管线
        // （已经在 GPUAccelerator::new() 中完成了）
        
        // Swift 第 288 行：初始化 PR 值
        let mut pr = initial_pr.to_vec();
        let vertex_count = csr.vertex_count as usize;
        
        // Swift 第 292-300 行：构建反向邻接表
        let mut reverse_adjacency = vec![Vec::new(); vertex_count];
        for u in 0..vertex_count {
            let start = csr.offsets[u] as usize;
            let end = csr.offsets[u + 1] as usize;
            for i in start..end {
                let v = csr.targets[i] as usize;
                reverse_adjacency[v].push(u as u32);
            }
        }
        
        // Swift 第 303-304 行：初始化受影响顶点集合
        let mut affected_set: HashSet<u32> = affected_vertices.iter().cloned().collect();
        let mut iteration = 0;
        
        // Swift 第 306 行：开始计时（可选）
        let start_time = std::time::Instant::now();
        
        // Swift 第 308 行：主循环
        while !affected_set.is_empty() && iteration < self.max_iterations {
            iteration += 1;
            
            // Swift 第 310 行：转换受影响的顶点为数组
            let affected_array: Vec<u32> = affected_set.iter().cloned().collect();
            let affected_count = affected_array.len() as u32;
            
            // Swift 第 314 行：复制当前 PR 值
            let mut new_pr = pr.clone();
            
            // Swift 第 316-321 行：创建 GPU 缓冲区
            // 调用 GPU 计算受影响顶点的 PR 值
            let updated_pr = self.gpu.compute_incremental_pagerank(
                &csr.offsets,
                &csr.targets,
                &pr,
                &affected_array,
                csr.vertex_count,
            );
            
            // Swift 第 345-354 行：检查收敛，找出新的受影响顶点
            let mut new_affected = HashSet::new();
            for i in 0..vertex_count {
                let diff = (updated_pr[i] - pr[i]).abs();
                if diff > self.tolerance {
                    // Swift 第 350-352 行：传播给邻居
                    for &neighbor in &reverse_adjacency[i] {
                        new_affected.insert(neighbor);
                    }
                }
            }
            
            // Swift 第 356-358 行：更新
            pr = updated_pr;
            affected_set = new_affected;
            
            // Swift 第 360-362 行：打印进度
            if iteration % 5 == 0 {
                let pr_sum: f32 = pr.iter().sum();
                println!("    增量迭代 {}：{} 个受影响顶点，PR 值之和 = {:.6}", 
                         iteration, affected_set.len(), pr_sum);
            }
        }
        
        // Swift 第 365-366 行：打印耗时
        let elapsed = start_time.elapsed();
        println!("    增量 PageRank：{} 次迭代，{:.2} ms", 
                 iteration, elapsed.as_secs_f64() * 1000.0);
        
        // Swift 第 368 行：返回结果
        pr
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use rand::Rng;
    
    #[test]
    fn test_incremental_pagerank() {
        // 检查是否有 GPU
        let incremental_pr = IncrementalPageRank::new();
        if incremental_pr.is_err() {
            println!("⚠️  没有 GPU，跳过测试");
            return;
        }
        let incremental_pr = incremental_pr.unwrap();
        
        // 创建一个简单的 CSR 图
        // 5 个顶点：0->1, 0->2, 1->2, 2->0, 3->4
        let mut csr = CSRGraph::new();
        
        // 添加顶点
        for i in 0..5 {
            let mut props = std::collections::HashMap::new();
            props.insert("id".to_string(), crate::csr_graph::PropertyValue::Int(i as i64));
            csr.add_vertex(i, props);
        }
        
        // 添加边
        let edges = vec![(0, 1), (0, 2), (1, 2), (2, 0), (3, 4)];
        csr.build_csr(&edges);
        
        // 初始 PR 值
        let initial_pr = vec![1.0 / 5.0; 5];
        
        // 受影响顶点：0, 1
        let affected_vertices = vec![0, 1];
        
        // 计算增量 PageRank
        let new_pr = incremental_pr.compute(&csr, &initial_pr, &affected_vertices);
        
        // 验证 PR 值之和接近 1.0
        let sum: f32 = new_pr.iter().sum();
        println!("PR 值之和：{}", sum);
        
        assert!((sum - 1.0).abs() < 0.01, "PR 值之和应该接近 1.0");
        
        println!("✅ 增量 PageRank 测试通过！");
    }
}
